File: src/binarySearch/InstallRouter.java
package binarySearch;

import java.io.*;
import java.util.Arrays;

public class InstallRouter {
    static int n;
    static int c;
    static int[] arr;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        String[] nc = br.readLine().split(" ");
        n = Integer.parseInt(nc[0]);
        c = Integer.parseInt(nc[1]);

        arr = new int[n];

        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }

        Arrays.sort(arr);

        long min = 1, max = arr[n - 1];

        while(min <= max){
            long mid = (min + max) / 2;

            if(isPossible(mid) < c) max = mid - 1;
            else min = mid + 1;
        }

        bw.write(String.valueOf(min - 1));
        bw.flush();
        bw.close();
    }

    private static int isPossible(long distance){
        int count = 1;
        int curIndex = 0;

        for (int i = 1; i < n; i++) {
            if(arr[i] - arr[curIndex] >= distance){
                count++;
                curIndex = i;
            }
        }

        return count;
    }
}

File: JS/binarySearch/installRouter.js
"use strict";

const fs = require("fs");

const input =
  process.platform === "linux"
    ? fs.readFileSync("/dev/stdin").toString().split("\n")
    : `5 3
1
2
8
4
9`.split("\n");

const [n, c] = input[0].split(" ").map((v) => parseInt(v));
const arr = input
  .slice(1)
  .map((v) => parseInt(v))
  .sort((a, b) => a - b);

let min = 1;
let max = arr[n - 1];

function isPossible(distance) {
  let count = 1;
  let curIndex = 0;

  for (let i = 1; i < n; i++) {
    if (arr[i] - arr[curIndex] >= distance) {
      count++;
      curIndex = i;
    }
  }

  return count;
}

while (min <= max) {
  let mid = parseInt((min + max) / 2);

  if (isPossible(mid) < c) max = mid - 1;
  else min = mid + 1;
}

console.log(min - 1);
